org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectCache.java

```diff
@@ -52,59 +55,66 @@
 @Component(service = MavenProjectCache.class)
 public class MavenProjectCache {
 
-  private static final int MAX_CACHE_SIZE = Integer.getInteger("m2e.project.cache.size", 20);
+  private static final int MAX_CACHE_SIZE = Integer.getInteger("m2e.project.cache.size", 50);
 
   private static final String CTX_MAVENPROJECTS = MavenProjectCache.class.getName() + "/mavenProjects";
 
   @Reference
   PlexusContainerManager containerManager;
 
-  private Cache<IMavenProjectFacade, MavenProject> mavenProjectCache;
+  private LoadingCache<CacheKey, CacheLine> loadingCache;
 
 
   /**
    * 
    */
   public MavenProjectCache() {
-    this.mavenProjectCache = CacheBuilder.newBuilder() //
+    this.loadingCache = CacheBuilder.newBuilder() //
         .maximumSize(MAX_CACHE_SIZE) //
-        .removalListener((RemovalNotification<IMavenProjectFacade, MavenProject> removed) -> {
+        .removalListener((RemovalNotification<CacheKey, CacheLine> removed) -> {
           Map<IMavenProjectFacade, MavenProject> contextProjects = getContextProjectMap();
-          IMavenProjectFacade facade = removed.getKey();
-          if(!contextProjects.containsKey(facade)) {
-            flushMavenCaches(facade.getPomFile(), facade.getArtifactKey(), false);
+          removed.getValue().projects.values().forEach(mavenProject -> {
+            if(!contextProjects.containsValue(mavenProject)) {
+              flushMavenCaches(mavenProject.getFile(), removed.getKey().artifactKey(), false);
+            }
+          });
+        }).build(new CacheLoader<CacheKey, CacheLine>() {
+
+          @Override
+          public CacheLine load(CacheKey key) throws Exception {
+            return new CacheLine();
           }
-        }).build();
+        });
   }
 
   /**
    * @param facade
    */
   public void invalidateProjectFacade(IMavenProjectFacade facade) {
-    mavenProjectCache.invalidate(facade);
+    CacheLine cacheLine = loadingCache.getIfPresent(new CacheKey(facade.getArtifactKey(), facade.getConfiguration()));
+    if(cacheLine != null) {
+      cacheLine.remove(facade.getPomFile());
```

Comment:
We can simply let the cache manage this, if the items is no longer in use it will be evicted some time later, or it can be reused iv become important later on.

Reply:
But the cache is size limited, so it could happen that another not-empty CacheLine is evicted instead of a now empty one reducing the effective size of the cache.